das/daser.go

```diff
@@ -42,28 +59,54 @@ func (d *DASer) Start(context.Context) error {
 		return err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	go d.sampling(ctx, sub)
+	// load latest DASed checkpoint
+	checkpoint, err := loadCheckpoint(d.ds)
+	if err != nil {
+		return err
+	}
+	log.Infow("loaded latest DASed checkpoint", "height", checkpoint)
+
+	dasCtx, cancel := context.WithCancel(context.Background())
+
+	go d.sample(dasCtx, sub, checkpoint)
+
 	d.cancel = cancel
 	return nil
 }
 
 // Stop stops sampling.
 func (d *DASer) Stop(ctx context.Context) error {
 	d.cancel()
-	select {
-	case <-d.done:
-		d.cancel = nil
-		return nil
-	case <-ctx.Done():
-		return ctx.Err()
+	// wait for both sampling routines to exit
+	for i := 0; i < 2; i++ {
+		select {
+		case <-d.catchUpDn:
+		case <-d.sampleDn:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
 	}
+	d.cancel = nil
+	return nil
 }
 
-// sampling validates availability for each Header received from header subscription.
-func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
-	defer sub.Cancel()
-	defer close(d.done)
+// sample validates availability for each Header received from header subscription.
+func (d *DASer) sample(ctx context.Context, sub header.Subscription, checkpoint int64) {
+	height := checkpoint
+
+	defer func() {
+		// store latest DASed checkpoint to disk
+		// TODO @renaynay: what sample DASes [100:150] and
+		//  stores latest checkpoint to disk as network head (150)
+		// 	but catchUp routine has only sampled from [1:40] so there is a gap
+		//  missing from (40: 100)?
```

Comment:
this TODO should ideally be resolved in this PR. I need to think about it a bit more because it might require a more hefty solution that would make this way less light-weight 😞  I need to think about it.

Reply:
@renaynay, we discussed on our call that https://github.com/celestiaorg/celestia-node/issues/180 should fix this TODO entirely, but in fact, it should only prevent re-sampling of already sampled heights and if it's implemented, the issue with the gap will still present.

The naive solution to the gap issue would be to store a checkpoint in a single `catchUp` routine that is described in the comment above. Basing on your example, when the `catchUp` is stopped it stores the height 40, so then DASer will start from 41 and resample [100:150] again, which we would like to avoid and where https://github.com/celestiaorg/celestia-node/issues/180 can help us.